### PR TITLE
 API Libraries > Polkadot API (PAPI)

### PR DESCRIPTION
**Pull request by nhussein11**
This page aims to add the PAPI library

### CODE REVIEW

File: develop/toolkit/api-libraries/papi.md [hunk 1]
```diff
@@ -1,4 +1,144 @@
...
+
+- To read a value from the **storage**, you can use the following pattern:
+
+    ```typescript
+    const result = await typedApi.query.Pallet.Query.getValue(
+        arg1,
+        arg2,
+        arg3,
+        { at: "best" }
+    )
```
Comment (0xLucca) [inline develop/toolkit/api-libraries/papi.md:95]: The example could be clearer. Replace with a specific storage read.

File: develop/toolkit/api-libraries/papi.md [hunk 2]
```diff
@@ -1,4 +1,144 @@
...
+
+With the `TypedApi`, you can read data from the blockchain using the following patterns:
+
+- To read a **constant** of the blockchain, you can use the `constants` property and access the constant you need:
+
+    ```typescript
+    const version = await typedApi.constants.System.Version()
+    ```
+
+- To read a value from the **storage**, you can use the following pattern:
```
Comment (0xLucca) [inline develop/toolkit/api-libraries/papi.md:87]:
```suggestion
- To read a value from the **storage**, you can use the `query` interface with the following pattern:
```

File: develop/toolkit/api-libraries/papi.md [hunk 3]
```diff
@@ -1,4 +1,144 @@
...
+    ```typescript
+    const result = await typedApi.query.Pallet.Query.getValue(
+        arg1,
+        arg2,
+        arg3,
+        { at: "best" }
+    )
+    ```
+
+- To directly query a **Runtime API**, you can use the `query` property and access the API you need, for example:
```
Comment (0xLucca) [inline develop/toolkit/api-libraries/papi.md:98]:
```suggestion
- To directly query a **Runtime API**, you can use the `apis` interface and access the API you need, for example:
```

File: develop/toolkit/api-libraries/papi.md [hunk 4]
```diff
@@ -1,4 +1,144 @@
...
+    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-wss.ts"
+    ```
+
+Now that you have set up the client, you can interact with the chain by reading and sending transactions.
+
+### Reading Chain Data
+
+With the `TypedApi`, you can read data from the blockchain using the following patterns:
+
+- To read a **constant** of the blockchain, you can use the `constants` property and access the constant you need:
```
Comment (0xLucca) [inline develop/toolkit/api-libraries/papi.md:81]:
```suggestion
- To read a **constant** of the blockchain, you can use the `constants` interface and access the constant you need:
```

File: develop/toolkit/api-libraries/papi.md [hunk 5]
```diff
@@ -1,4 +1,144 @@
...
+    ``` 
+
+    For instance, to execute a transaction using the call data, you can use the following snippet:
+
+    ```typescript
+    const callData = Binary.fromHex("0x00002470617065726d6f6f6e")
+    const tx: Transaction = typedApi.txFromCallData(callData)
+    ```
+
+    Ensure you replace `INSERT_CALL_DATA` with the actual call data, a hex string.
```
Comment (0xLucca) [inline develop/toolkit/api-libraries/papi.md:138]: There is no INSERT_CALL_DATA in the snippet

File: .snippets/code/develop/toolkit/api-libraries/papi/send-tx-transfer-keep-alive.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { MultiAddress } from '@polkadot-api/descriptors';
+
+const tx: Transaction = typedApi.tx.Balances.transfer_keep_alive({
+  dest: MultiAddress.Id('INSERT_DESTINATION_ADDRESS'),
+  value: BigInt(INSERT_VALUE),
+});
```
Comment (kianenigma) [inline .snippets/code/develop/toolkit/api-libraries/papi/send-tx-transfer-keep-alive.ts:6]: Has there been an issue or a roadmap item to make the code snippets in this repo compile?

**Review by eshaben — COMMENTED** [review]
Nice work!

File: develop/toolkit/api-libraries/papi.md [hunk 7]
```diff
@@ -1,4 +1,140 @@
...
+    ```
+
+Then, obtain the latest metadata from the target chain and generate the necessary types:
+
+```bash
+# Add the target chain
+npx papi add dot -n polkadot
+```
+
+The `papi add` command initializes papi add, assigns the chain a custom name and specifies downloading metadata from the Polkadot chain. You can replace `dot` with the chain name if you want to add a different chain. Once the latest metadata is downloaded, generate the required types:
```
[the remark this answers is not in the document]
Comment (nhussein11) [reply develop/toolkit/api-libraries/papi.md]: You are right; it's confusing. I paraphrased it in [20bc1f5](https://github.com/polkadot-developers/polkadot-docs/pull/138/commits/20bc1f55e7174a1d9a4ebb54b329389955d05571)

File: develop/toolkit/api-libraries/papi.md [hunk 8]
```diff
@@ -1,4 +1,140 @@
...
+    --8<-- "code/develop/toolkit/api-libraries/papi/api-instantiation-wss.ts"
+    ```
+
+Now that you have set up the client, you can interact with the chain by reading and sending transactions.
+
+### Reading Chain Data
+
+With the `TypedApi`, you can read data from the blockchain using the following patterns:
+
+- To read a **constant** of the blockchain, you can use the `constants` interface and access the constant you need:
```
[the remark this answers is not in the document]
Comment (nhussein11) [reply develop/toolkit/api-libraries/papi.md:81]: it was just to emphasize the "type" of query or request we are doing to the chain, like you can either query a `constant`, `storage` portions of the chain OR do queries via `runtime apis`

image

@eshaben, do you believe there is another way to be more explicit in achieving the same? I'm not 100% convinced with this approach, but it's the only one that came to my mind tbh

[the remark this answers is not in the document]
Comment (eshaben) [reply develop/toolkit/api-libraries/papi.md:81]: Maybe start the list item with each one:

```
The `TypedApi` provides a streamlined way to read blockchain data through three main interfaces, each designed for specific data access patterns:

- **Constants** - access fixed values or configurations on the blockchain using the `constants` interface:

    ```typescript
    const version = await typedApi.constants.System.Version();
    ```

- **Storage queries** - retrieve stored values by querying the blockchain’s storage via the `query` interface:

    ```typescript
    const asset = await api.query.ForeignAssets.Asset.getValue(
      token.location,
      { at: 'best' },
    );
    ```

- **Runtime APIs** - interact directly with runtime APIs using the `apis` interface:

    ```typescript
    const metadata = await typedApi.apis.Metadata.metadata();
    ```
```

File: develop/toolkit/api-libraries/papi.md [hunk 9]
```diff
@@ -1,4 +1,140 @@
...
+    ```typescript
+    const callData = Binary.fromHex('0x00002470617065726d6f6f6e');
+    const tx: Transaction = typedApi.txFromCallData(callData);
+    ```
+
+For more information about sending transactions, refer to the [Transactions](https://papi.how/typed/tx#transactions){target=\_blank} page.
+
+## Where to Go Next
+
+For an in-depth guide on how to use PAPI, refer to the official [PAPI](https://papi.how/typed){target=\_blank} documentation.
```
Comment (eshaben) [inline develop/toolkit/api-libraries/papi.md:140]:
```suggestion
For an in-depth guide on how to use PAPI, refer to the official [PAPI](https://papi.how/){target=\_blank} documentation.
```

File: develop/toolkit/api-libraries/papi.md [hunk 10]
```diff
@@ -1,4 +1,140 @@
...
+    );
+    ```
+
+- To directly query a **Runtime API**, you can use the `apis` interface and access the API you need, for example:
+
+    ```typescript
+    const metadata = await typedApi.apis.Metadata.metadata();
+    ```
+
+To learn more about the different actions you can perform with the `TypedApi`, refer to the [Typed API Reference](https://papi.how/typed){target=\_blank}.
```
Comment (eshaben) [inline develop/toolkit/api-libraries/papi.md:102]:
```suggestion
To learn more about the different actions you can perform with the `TypedApi`, refer to the [TypedApi reference](https://papi.how/typed){target=\_blank}.
```

File: develop/toolkit/api-libraries/papi.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (nhussein11) [reply develop/toolkit/api-libraries/papi.md:81]: makes sense, it's much clearer. Ty!

**Review by eshaben — APPROVED** [review]
LGTM!

**Pull request merged**